### PR TITLE
Handle table aliases correctly

### LINKED ISSUE #14: Alias removal is over-eager; removes tables that are shadowed
```
(deftest tables-with-complex-aliases-test
  (testing "With an alias that is also a table name"
    (is (= #{"user" "user2_final"}
           (tables
            "SELECT legacy_user.id AS old_id,
                    user.id AS new_id
             FROM user AS legacy_user
             OUTER JOIN user2_final AS user
             ON legacy_user.email = user2_final.email;")))))
```

### PR DESCRIPTION
**Pull request by tsmacdonald**
[Fixes #14]

That was obvious; I should've done this weeks ago.

### CODE REVIEW

**Comment by tsmacdonald** [comment]
image

hey hey

File: java/com/metabase/macaw/AstWalker.java [hunk 1]
```diff
@@ -386,21 +392,11 @@ public void visit(OverlapsCondition overlapsCondition) {
...
-        //
-        // select orders.id from orders;
-        //
-        // you need to visit the `orders` in `orders.id`. That table is distinct from the one in `from orders`.
         if (tableColumn.getTable() != null
                 && tableColumn.getTable().getName() != null) {
-            visit(tableColumn.getTable());
+            // visiting aliases (e.g., the `o` in `o.id` in `select o.id from orders o`) is unhelpful if we're trying
+            // to get the set of actual table names used. However, for query rewriting it's necessary
+            visitAlias(tableColumn.getTable());
```
Comment (crisptrutski) [inline java/com/metabase/macaw/AstWalker.java:399]: I think it's a potential gotcha that referencing non-aliases tables directly will result on them getting hit with this callback. Perhaps there's another name we could use that avoid this, maybe `visitReference(table)`?

In either case I think it's worth adding a test that shows how this works:

```clojure
    (is (= #{"user" "loser"}
           (references
            "SELECT 
                    user.id AS user_id,
                    loser.id AS loser_id
             FROM user
             OUTER JOIN user AS loser
             ON user.email = user.bully_email;")))))
```

If we find another word that works, we can then define `aliases` as `(remove (tables) (references))` ✨

**Comment by tsmacdonald** [comment]
@crisptrutski I'm open to bike-shedding, but I opted for the wordy-but-obvious "column-specifier". "reference" and its ilk felt very overloaded

**Comment by tsmacdonald** [comment]
image


Awww…

(but adding tests is a good thing!)

**Comment by crisptrutski** [comment]
> I opted for the wordy-but-obvious "column-specifier".

I don't mind wordy, but I find this name confusing still - I'd expect it to refer to the entire expression, even things like selecting constants.

How about "column-qualifier"?

**Comment by crisptrutski** [comment]
I asked ChatGPT and it preferred "table-qualifier"

File: src/macaw/rewrite.clj [hunk 2]
```diff
@@ -66,13 +66,18 @@
...
   "Given a SQL query and its corresponding (untransformed) AST, apply the given table and column renames."
   [sql parsed-ast {table-renames :tables, column-renames :columns}]
   (-> parsed-ast
       (mw/walk-query
-       {:table  (fn [^Table table] (when-let [name' (get table-renames (.getName table))]
-                                     (.setName table name')))
-        :column (fn [^Column column] (when-let [name' (get column-renames (.getColumnName column))]
-                                       (.setColumnName column name')))})
+       {:table            (partial rename-table table-renames)
+        :column-qualifier (partial rename-table table-renames)
```
Comment (crisptrutski) [inline src/macaw/rewrite.clj:80]: Don't we need to make sure that the qualifier isn't an alias that shadows the table name here? Annoying and unlikely as it is.

**Review by crisptrutski — APPROVED** [review]
I think there's a subtle bug in here but his is safer than before so not gonna block on it.

**Pull request merged**